File: python/mirage/commands/spec/flag_view.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Mapping
from typing import Generic, TypeVar

from mirage.commands.spec.constants import flag_kwarg_name
from mirage.commands.spec.types import CommandSpec, FlagValue
from mirage.types import PathSpec

T = TypeVar("T")


class FlagBag(dict[str, T], Generic[T]):
    """Flag values with a separate tape of occurrences in scan order.

    Args:
        values (Mapping[str, T] | None): Values to copy, preserving their tape.
    """

    def __init__(self, values: Mapping[str, T] | None = None) -> None:
        super().__init__(values or {})
        self.occurrences: list[tuple[str, str | bool | int]] = (list(
            values.occurrences) if isinstance(values, FlagBag) else [])


class FlagView:
    """Typed read-only view over raw flag kwargs.

    Commands receive flags as an untyped mapping from the dispatcher; this
    view is the one sanctioned way to read them, replacing ad-hoc
    `flags.get(...) is True` and isinstance chains.

    Args:
        flags (Mapping[str, FlagValue] | None): raw flag kwargs.
        spec (CommandSpec | None): when given, reads of names the spec does
            not declare raise KeyError. A missing key is otherwise
            indistinguishable from "flag not passed", so a typo in the name
            would silently read as False/None.
    """

    def __init__(self,
                 flags: Mapping[str, FlagValue] | None,
                 spec: CommandSpec | None = None) -> None:
        self._flags = flags if flags is not None else {}
        self._allowed = spec_flag_names(spec) if spec is not None else None

    def _key(self, name: str) -> str:
        if self._allowed is not None and name not in self._allowed:
            raise KeyError(f"flag {name!r} is not declared by the command "
                           f"spec (known: {sorted(self._allowed)})")
        return name

    def typed_order(self, *names: str) -> list[str]:
        """The given flag names, ordered by their recorded occurrences.

        The parser fills the bag in scan order and every hop between
        (kwargs, dict copies) preserves insertion order, so a key's
        position is its last occurrence for a scalar, or its first for an
        accumulating option; a flag supplied
        by a default or the environment lands after every typed one.
        Names the line never carried are dropped. This is what an
        order-sensitive option family (grep's --include/--exclude,
        where the later kind overrides the earlier) reads, since the
        bag has no per-occurrence positions.

        Args:
            names (str): flag names to order.
        """
        wanted = {self._key(n) for n in names}
        return [k for k in self._flags if k in wanted]

    def occurrences(self, *names: str) -> list[tuple[str, FlagValue]]:
        """Read each typed occurrence, then defaults without a typed value.

        Args:
            names (str): Spec-bound option names to read.
        """
        wanted = {self._key(name) for name in names}
        tape = self._flags.occurrences if isinstance(self._flags,
                                                     FlagBag) else []
        result: list[tuple[str, FlagValue]] = [
            (name, value) for name, value in tape
            if name in wanted and name in self._flags
        ]
        seen = {name for name, _ in result}
        for name in self.typed_order(*names):
            if name not in seen:
                value = self._flags[name]
                if isinstance(value, list):
                    result.extend((name, item) for item in value)
                else:
                    result.append((name, value))
        return result

    def as_bool(self, name: str) -> bool:
        value = self._flags.get(self._key(name))
        if isinstance(value, bool):
            return value
        # A count flag holds an int; any occurrence reads as set.
        return isinstance(value, int) and value > 0

    def as_int(self, name: str) -> int | None:
        value = self._flags.get(self._key(name))
        if isinstance(value, bool):
            return None
        if isinstance(value, int):
            return value
        if not isinstance(value, str):
            return None
        try:
            return int(value)
        except ValueError as exc:
            raise ValueError(f"flag '{name}' expects an integer, "
                             f"got '{value}'") from exc

    def as_float(self, name: str) -> float | None:
        value = self._flags.get(self._key(name))
        if isinstance(value, bool):
            return None
        if isinstance(value, (int, float)):
            return float(value)
        if not isinstance(value, str):
            return None
        try:
            return float(value)
        except ValueError as exc:
            raise ValueError(f"flag '{name}' expects a number, "
                             f"got '{value}'") from exc

    def as_str(self, name: str) -> str | None:
        value = self._flags.get(self._key(name))
        return value if isinstance(value, str) else None

    def as_list(self, name: str) -> list[str]:
        value = self._flags.get(self._key(name))
        if isinstance(value, list):
            return [item for item in value if isinstance(item, str)]
        if isinstance(value, str):
            return [value]
        return []

    def as_paths(self, name: str) -> list[PathSpec]:
        # PATH-typed flag values arrive as PathSpec in the python
        # executor; the TypeScript flag bag carries their resolved
        # virtual-path strings, so its counterpart is asList.
        value = self._flags.get(self._key(name))
        if isinstance(value, list):
            return [item for item in value if isinstance(item, PathSpec)]
        if isinstance(value, PathSpec):
            return [value]
        return []

    def raw(self, name: str) -> FlagValue | None:
        return self._flags.get(self._key(name))


def spec_flag_names(spec: CommandSpec) -> frozenset[str]:
    """Collect the kwarg names a spec's options can produce.

    One name per option: the long spelling when an option declares
    both, matching the parser's canonical dest. Keeping the short
    spelling here too would let a stale ``fl.as_bool("a")`` stay legal
    and read False forever after dest unification; canonical-only
    turns that silent miss into a KeyError.

    Args:
        spec (CommandSpec): command spec whose options to enumerate.
    """
    names: set[str] = set()
    for option in spec.options:
        canonical = option.long if option.long is not None else option.short
        if canonical is not None:
            names.add(flag_kwarg_name(canonical))
    return frozenset(names)

File: typescript/packages/core/src/commands/spec/flag_view.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { flagKwargName } from './constants.ts'
import { compareCodePoints } from '../../utils/sort.ts'
import type { CommandSpec, FlagValue } from './types.ts'

const occurrenceTapes = new WeakMap<Record<string, FlagValue>, [string, FlagValue][]>()

export function flagOccurrences(flags: Record<string, FlagValue>): [string, FlagValue][] {
  let tape = occurrenceTapes.get(flags)
  if (tape === undefined) {
    tape = []
    occurrenceTapes.set(flags, tape)
  }
  return tape
}

/**
 * Collect the kwarg names a spec's options can produce.
 *
 * One name per option: the long spelling when an option declares both,
 * matching the parser's canonical dest. Keeping the short spelling here
 * too would let a stale `fl.asBool('a')` stay legal and read false
 * forever after dest unification; canonical-only turns that silent miss
 * into a throw. Mirrors Python's `spec_flag_names`.
 */
export function specFlagNames(spec: CommandSpec): ReadonlySet<string> {
  const names = new Set<string>()
  for (const option of spec.options) {
    const canonical = option.long ?? option.short
    if (canonical !== null) names.add(flagKwargName(canonical))
  }
  return names
}

/**
 * Typed read-only view over raw flag kwargs.
 *
 * Commands receive flags as an untyped record from the dispatcher; this
 * view is the one sanctioned way to read them, replacing ad-hoc
 * `flags.x === true` checks and typeof chains. Mirrors Python's
 * `FlagView`.
 *
 * When constructed with a spec, reading a name the spec does not declare
 * throws. A missing key is otherwise indistinguishable from "flag not
 * passed", so a typo in the name would silently read as false/undefined.
 */
export class FlagView {
  private readonly flags: Readonly<Record<string, FlagValue>>
  private readonly allowed: ReadonlySet<string> | null

  constructor(flags?: Readonly<Record<string, FlagValue>>, spec?: CommandSpec) {
    this.flags = flags ?? {}
    this.allowed = spec === undefined ? null : specFlagNames(spec)
  }

  private key(name: string): string {
    if (this.allowed !== null && !this.allowed.has(name)) {
      throw new Error(
        `flag '${name}' is not declared by the command spec ` +
          `(known: ${[...this.allowed].sort(compareCodePoints).join(', ')})`,
      )
    }
    return name
  }

  /**
   * The given flag names, ordered by their recorded occurrences.
   *
   * The parser fills the bag in scan order and every hop between
   * (object spreads, copies) preserves string-key insertion order, so
   * a key's position is its last occurrence for scalars, first for accumulating options; a flag
   * supplied by a default or the environment lands after every typed
   * one. Names the line never carried are dropped. This is what an
   * order-sensitive option family (grep's --include/--exclude, where
   * the later kind overrides the earlier) reads, since the bag has no
   * per-occurrence positions.
   */
  typedOrder(...names: string[]): string[] {
    const wanted = new Set(names.map((n) => this.key(n)))
    return Object.keys(this.flags).filter((k) => wanted.has(k))
  }

  occurrences(...names: string[]): [string, FlagValue][] {
    const wanted = new Set(names.map((name) => this.key(name)))
    const result = flagOccurrences(this.flags).filter(
      ([name]) => wanted.has(name) && name in this.flags,
    )
    const seen = new Set(result.map(([name]) => name))
    for (const name of this.typedOrder(...names)) {
      if (seen.has(name)) continue
      const value = this.flags[name]
      if (value === undefined) continue
      for (const item of Array.isArray(value) ? value : [value]) result.push([name, item])
    }
    return result
  }

  asBool(name: string): boolean {
    const value = this.flags[this.key(name)]
    if (typeof value === 'boolean') return value
    // A count flag holds a number; any occurrence reads as set.
    return typeof value === 'number' && value > 0
  }

  asInt(name: string): number | undefined {
    const value = this.flags[this.key(name)]
    if (typeof value === 'number') return value
    if (typeof value !== 'string') return undefined
    // Python's int() is all-or-nothing: it accepts surrounding whitespace
    // and underscore separators and raises on anything else. parseInt would
    // instead take the numeric prefix of '5x' and hand back NaN for 'abc',
    // and NaN still satisfies `number`, so a bad value would flow onward as
    // a number rather than being rejected.
    const text = value.trim()
    if (!/^[+-]?\d+(?:_\d+)*$/.test(text)) {
      throw new Error(`flag '${name}' expects an integer, got '${value}'`)
    }
    return Number.parseInt(text.replaceAll('_', ''), 10)
  }

  asFloat(name: string): number | undefined {
    const value = this.flags[this.key(name)]
    if (typeof value === 'number') return value
    if (typeof value !== 'string') return undefined
    // All-or-nothing like Python's float(), mirroring asInt: parseFloat
    // would take the numeric prefix of '2.5x' and hand back NaN for
    // 'abc', and NaN still satisfies `number`.
    const text = value.trim()
    if (
      !/^[+-]?(?:\d+(?:_\d+)*(?:\.(?:\d+(?:_\d+)*)?)?|\.\d+(?:_\d+)*)(?:[eE][+-]?\d+(?:_\d+)*)?$/.test(
        text,
      )
    ) {
      throw new Error(`flag '${name}' expects a number, got '${value}'`)
    }
    return Number.parseFloat(text.replaceAll('_', ''))
  }

  asStr(name: string): string | undefined {
    const value = this.flags[this.key(name)]
    return typeof value === 'string' ? value : undefined
  }

  asList(name: string): string[] {
    const value = this.flags[this.key(name)]
    if (Array.isArray(value)) return value.filter((v) => typeof v === 'string')
    if (typeof value === 'string') return [value]
    return []
  }

  raw(name: string): FlagValue | undefined {
    return this.flags[this.key(name)]
  }
}
